src/main/java/com/ibm/nfvodriver/service/AuthenticatedRestTemplateService.java

```diff
@@ -138,19 +151,70 @@ private RestTemplate getBasicAuthenticatedRestTemplate(final Map<String, String>
-        if (StringUtils.hasText(authenticationProperties.get(AUTHENTICATION_SCOPE))) {
-            resourceDetails.setScope(Arrays.asList(authenticationProperties.get(AUTHENTICATION_SCOPE).split(",")));
+    private static AuthorizationGrantType mapStringToGrantType(String grantTypeString, AuthorizationGrantType defaultGrantType) {
+        switch (grantTypeString) {
+            case "client_credentials":
+                return AuthorizationGrantType.CLIENT_CREDENTIALS;
+            case "authorization_code":
+                return AuthorizationGrantType.AUTHORIZATION_CODE;
+            case "password":
+                return AuthorizationGrantType.PASSWORD;
+            case "refresh_token":
+                return AuthorizationGrantType.REFRESH_TOKEN;
+            default:
+                return defaultGrantType;
+        }
+    }
+    private static AuthorizationGrantType getOrDefaultForGrantType(final Map<String, String> properties, String grantTypeKey, AuthorizationGrantType defaultGrantType) {
+
+        String grantTypeStr = properties.get(grantTypeKey);
+        if (grantTypeStr == null) {
+            return defaultGrantType;
         }
 
-        logger.info("Configuring OAuth2 authenticated RestTemplate.");
-        return restTemplateBuilder.configure(new OAuth2RestTemplate(resourceDetails));
+        return mapStringToGrantType(grantTypeStr, defaultGrantType);
     }
+    
+    private RestTemplate getOAuth2AuthenticatedRestTemplate(final Map<String, String> authenticationProperties) {
+        ClientRegistration.Builder clientRegistrationBuilder = ClientRegistration.withRegistrationId(authenticationProperties.get(AUTHENTICATION_CLIENT_ID))
+                .clientId(authenticationProperties.get(AUTHENTICATION_CLIENT_ID))
+                .clientSecret(authenticationProperties.get(AUTHENTICATION_CLIENT_SECRET))
+                .authorizationGrantType(getOrDefaultForGrantType(authenticationProperties, AUTHENTICATION_GRANT_TYPE, AuthorizationGrantType.CLIENT_CREDENTIALS ))
+                .tokenUri(authenticationProperties.get(AUTHENTICATION_ACCESS_TOKEN_URI));
+
+        if (StringUtils.hasText(authenticationProperties.get(AUTHENTICATION_SCOPE))) {
+                    clientRegistrationBuilder.scope(Arrays.asList(authenticationProperties.get(AUTHENTICATION_SCOPE).split(",")));
+                }  
+        ClientRegistration clientRegistration = clientRegistrationBuilder.build();
+        return restTemplateBuilder
+                .additionalInterceptors(new OAuthClientCredentialsRestTemplateInterceptor(authorizedClientManager(clientRegistration), clientRegistration))
+                .build();
+
+        }
+        
+    private OAuth2AuthorizedClientManager authorizedClientManager(ClientRegistration clientRegistration) {
+        var authorizedClientProvider = OAuth2AuthorizedClientProviderBuilder.builder()
+                .clientCredentials()
```

Comment:
clientCredentials are hardcoded here. Could you try to remove this line and proceed testing.

Reply:
for authorizedClientProvider Grant_type is  required parameter for the builder  and its expects us to define at least one authorization grant type. If we don't specify any grant type (like .clientCredentials(), .authorizationCode(), .password(), etc.), it will result  in error stating "authorizedClientProviders cannot be empty."
<img width="1075" alt="Screenshot 2024-02-06 at 2 09 42 PM" src="https://github.com/IBM/sol005-lifecycle-driver/assets/155344158/c841181a-6cc3-4440-a577-4bc6643145d0">

Reply:
Could you take the value from deployment location and apply the appropriate method here.

Reply:
done